### ISSUE TITLE
Docking client not rendering on drag since 13.4.9-canary.3

### ISSUE THREAD

**Issue by aranselmo** [vercel/next.js#54512]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
next-js-server-rc-dock]$ npx --no-install next info

    Operating System:
      Platform: linux
      Arch: x64
      Version: #1 SMP Sat Jun 10 12:24:53 UTC 2023
    Binaries:
      Node: 18.9.1
      npm: 9.5.0
      Yarn: N/A
      pnpm: N/A
    Relevant Packages:
      next: 13.4.20-canary.4
      eslint-config-next: N/A
      react: 18.2.0
      react-dom: 18.2.0
      typescript: 5.1.6
    Next.js Config:
      output: N/A
```

### Which area(s) of Next.js are affected? (leave empty if unsure)

App Router

### Link to the code that reproduces this issue or a replay of the bug

https://github.com/aranselmo/next-js-server-rc-dock/tree/nonworking-with-v13.4.20-canary.4

### To Reproduce

Use an app like the above link with latest rc-dock  and latest next.js to run it. In the non-working branch you will not see drag snap to buttons on dragging a panel, whereas in the main working branch, you will see snap to buttons on dragging. The only difference is the  version of next.js used so this was introduced in the last few months of changes.

Really, all you theoretically need for this example is an import {DockLayout} from 'rc-dock' and page returning DockLayout object with one of the examples from https://ticlo.github.io/rc-dock/examples/.

This is just a realistic simplification of a docking with state stored in a context provider.

### Describe the Bug

Fundamentally the bug is the latest next.js isn't allowing a third-party client component to trigger a rerender of its root element on dragging.

On the above main branch using next.js **13.4.8** (which works as it used to work), the rc-dock allows dragging one tab panel over another panel and it will render a snap-to pop up in the center that lets you snap the panel above/below/etc. in relation to the drag target.

On the above branch using latest next.js, the rc-dock no longer renders the DockLayout root component on drag and so never shows the dock snaps and prevents rc-dock from working.

I tracked this down to a change in **13.4.9-canary.3** which mentions not rerendering the root element unnecessarily, but this was a breaking change for my apps at least and the only difference is updating next.js version.

### Expected Behavior

I think there needs to be more thought put into the change made in 13.4.9-canary.3 that prevents root rerendering and either make it some sort of flag that can be disabled/enabled or something.

Otherwise, I'd like a suggested workaround for how to wrap my DockLayout component that would allow it to work with the latest next.js.

As is, I've had to lock my application to version 13.4.9-canary.2 as that is the last working version for this combination.

### Which browser are you using? (if relevant)

Latest Chrome, but shouldn't matter

### How are you deploying your application? (if relevant)

npm run dev

**Comment by aranselmo** [comment]
So I've been able to narrow this down to the 6dafc4229d1bbcf11b9a5c243f39a8e76de5a27f commit between 13.4.9

> This is still occurring in the latest v13.5.4-canary.9 build and everything post 13.4.9-canary.2 so I'm locked at this version until this can be worked around.

So I've been able to narrow this down to the next.js https://github.com/vercel/next.js/commit/6dafc4229d1bbcf11b9a5c243f39a8e76de5a27f commit between 13.4.9-canary.2 and 13.4.9-canary.3. 

That commit updated react to new version, so I guess this issue is happening due to some incompatibility of rc-dock with next.js's react@18.3.0-canary-7118f5dd7-20230705 and above.

I was able to run and step into the rc-dock code and the only thing I can figure out is before that version of react, starting a drag would initiate a render() call that would update the panel with dock-panel-dropping class that would display the drop container. 

But after that update, the render() only gets called on initial render and drag never triggers it, which means there's no way to drag/drop the components. 

There's a change in react.development.js that might be it that looks like 

```
      var maybePromise = fn(); // If the function component returns a promise, it's likely an async
      // component, which we don't yet support. Attach a noop catch handler to
      // silence the error.
      // TODO: Implement component stacks for async client components?

      if (maybePromise && typeof maybePromise.catch === 'function') {
        maybePromise.catch(function () {});
      }
```
So I'm wondering if perhaps next.js does things in an async way that would hit that and cause certain actions to never trigger or something.
